quinn/src/delay_queue.rs

```diff
@@ -0,0 +1,521 @@
+use std::{fmt, ops::RangeInclusive};
+
+use slab::Slab;
+
+/// Stores values to be yielded at specific times in the future
+///
+/// Time is expressed as a bare u64 representing an absolute point in time. The caller may use any
+/// consistent unit, e.g. milliseconds, and any consistent definition of time zero. Larger units
+/// limit resolution but make `poll`ing over the same real time interval proportionately faster,
+/// whereas smaller units improve resolution, limit total range, and reduce `poll` performance.
+#[derive(Debug)]
+pub struct DelayQueue<T> {
+    /// Definitions of each active timer
+    timers: Slab<TimerState<T>>,
+
+    /// Hierarchical timer wheel
```

Comment:
So I need a bunch more/different documentation here. I've not encountered this data structure before, and I feel the documentation goes straight into a bunch details while skipping some of the big picture of how this data structure is supposed to work. One thing that seems mostly implicit here is the use of bits from the timer values as entries into the levels (which I only accidentally somewhat inferred from the `next_tick` docstring).

The structure with the `timers` field acting as a value map with `levels` acting as a hierarchical lookup table is also not explained very much here.

Is it fair to say there's one `slot` per bit in the `u64`, except organized in levels to make it a little faster to make the next one in case the slots are sparse? Maybe the documentation could show the expected organization in one example case with say two near and one far timer.

Reply:
I've expanded the documentation here substantially and added a diagram. Is it clearer now?

> Is it fair to say there's one slot per bit in the u64, except organized in levels to make it a little faster to make the next one in case the slots are sparse?

No, it is more or less coincidental that the bit width of a time is the same as the number of slots in a level. I'm cribbing tokio's guesstimate of an efficient level size here, and I suspect they're motivated by the use of a bitfield to represent slot occupancy to accelerate scans, which I'm leaving for future work here. The data structure will work for any value of `LOG_2_SLOTS`.

Reply:
> I've expanded the documentation here substantially and added a diagram. Is it clearer now?

It's much better! One nit with it is that the shift from base-10 to base-2 is a little misleading (and it might be easy to miss the shift towards the end of the docstring). Maybe it would be better to make the diagram base-2 as well, maybe in the `u8` domain or something?

Reply:
Moved that information to the front and filled it out into a better introductory paragraph. I like using base 10 because it's particularly easy to read/visualize, and levels with only 2 slots each might be a bit confusing, so hopefully this makes the compromise more obvious.